---
title: "MetPlast Tutorial"
author: "Lucio D'Andrea"
date: "8/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Plants as sessile organisms are unable to evade unfavorable growing conditions by simply moving away. Hence, they evolved a unique phenotypic plasticity that allows them to better adapt or survive to challenging environments. Metabolic plasticity is the ability of plants to biosynthesize a myriad of specialized compounds that allows them to cope with changes in their immediate surroundings. Thus, specialized metabolites are involved in a wide variety of ecological processes such as herbivorous attack, interaction with neighboring plants, as well as dealing with changes in light, or temperature conditions.

The evolution of plant metabolic plasticity has been mainly driven by gene duplication, (or even whole genome duplication) followed by neo-functionalization (REF). Gene duplication has been proven to shape the evolution of several specialize metabolic pathways, such as xxx and xxx. However, the effect of WGD on the metabolic plasticity remains to be elucidated. Possibly, the duplication of the whole genome allowed plants to screen a wider phenotypic space under stress conditions, promoting innovation, rapid adaptation and ultimately, speciation5. 

 Artificial selection processes have also influenced plant metabolic repertoire. Domestication, i.e., the process of selecting plants to increase their suitability to human requirements, as well as crop improvement has caused genetic bottlenecks and massive reduction of the allelic diversity. Thus, artificial selection has introduced quantitative changes in various nutrition compounds. For instance, studies on tomato domestication have shown a major reduction in the levels of the anti-nutritional steroidal glycoalkaloids in ripe fruits. Although, both natural and artificial selection have been pointed as major forces shaping the biosynthesis and accumulation of several specialized metabolites, the evaluation of their effects on the metabolome and metabolic plasticity are in their infancy.

Information theory provides a statistical framework that allows to quantify and evaluate metabolic plasticity (REF). Metabolome diversity and specialization can be calculated based on the Shannon entropy of the metabolic frequency distribution. Shannon entropy it a very useful parameter, that measures the information held in a set of data. Thus, its calculation can be used to estimate different parameters associated with a given metabolome: (1) Hj index, metabolome diversity; (2) j index, metabolic profile specialization; (3) Si index, metabolic specificity of individual metabolites. The individual calculation of these parameters was successfully applied on LC-MS/MS data to understand the dynamics of different plant species’ metabolomes (REF, don´t forget Moghe paper).

Here, we present MetDiv, an R-package that integrates the calculation, and visualization of Shannon-based metabolic plasticity parameters. We evaluate the effect of crop domestication by comparing the proposed parameters between the domesticated Solanum lycopersicum, the semi-domesticated S. lycopersicum var. cerisiforme and the wild relative S. pimpinellifolium.

# Upload external packages

This package has several external dependencies: 

1. dlookr
2. ggplot2
3. ggfortify
4. tidyverse
5. data.table
6. grid.Extra

Please, in case these packages are already available, install them using: 

```{r}
# install.packages()
```

Then, upload them:

```{r}

library("dlookr")

library("ggplot2")

library("ggfortify")

library ("tidyverse")

library("data.table")

library("gridExtra")

```

# Data 

## Uploading

When upload the Data set, the first column should be named "Compounds".

```{r, echo=TRUE}
Data <- read.csv2(file = "Test.csv", header = TRUE, row.names = "Compounds")

library(rmarkdown)
paged_table(head(Data))

```

## Replacing NAs values

The NAs values are replaced by the minimum value found in the data set divided by 1e6.

```{r}
Data[is.na(Data)] <- (min(Data, na.rm=TRUE))/1000000

library(rmarkdown)
paged_table(head(Data))
```

## Initial statistical analysis 

```{r}
describe(Data)
normality(Data)

Data.t <- t(Data)
pca_res <- prcomp(Data.t, scale=TRUE)
autoplot (pca_res, label=TRUE)

```

# Metabolic Parameters

## Upload "Test" Package

Install and upload the Test Package from devtools

```{r}
library(Test)

```

## Metabolic Plasticity (MetDiv function)

```{r}
Hj <- MetDiv (Data)
```

## Metabolite and Metabolome Specialization Index (MetSpec function)

```{r}
Dj <- MetSpec (Data)
```

## Metabolite Specialization Analysis (MetliteSpec function)

```{r}
MetliteSpec <- MetliteSpec(Data)
```

## Metabolic Plasticity Parameters Summary

This functions generates a table summarizing the Hj, number of peaks, and Dj parameters per species.

This function takes dataframe generated by MetDiv() and MetSpec().This data frame allows the pairwise comparision of the different calculated species paramenters

```{r}
MetPar <- MetPar(Data)

# Statistical Analysis 

## Extracting Parameters data frame

MetPar_df <- as.data.frame(MetPar[[1]])

## ANOVA analysis

library (ggpubr)

compare_means(Hj ~ Species, data = MetPar_df)

```

## Metabolic Plasticity Statistics

This functions generates a table summarizing the Hj, and the divergence associated to each Hj (HRj) and the Kullback–Leibler divergence (Divj). It takes dataframe generated by MetDiv().

```{r}
MetStats <- MetStats(Data)
```

